presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java

```diff
@@ -1223,6 +1251,77 @@ public void testBadUrlMissingSchema()
         }
     }
 
+    @Test(timeOut = 70000)
+    public void testQueryCancellation()
+            throws Exception
+    {
+        try (Connection connection = createConnection("blackhole", "blackhole");
+                Statement statement = connection.createStatement()) {
+            statement.executeUpdate("CREATE TABLE test_cancellation (key BIGINT) " +
+                    "WITH (" +
+                    "   split_count = 1, " +
+                    "   pages_per_split = 1, " +
+                    "   rows_per_page = 1, " +
+                    "   page_processing_delay = '1m'" +
+                    ")");
+        }
+
+        CountDownLatch queryStarted = new CountDownLatch(1);
+        CountDownLatch queryFinished = new CountDownLatch(1);
+        AtomicReference<String> queryId = new AtomicReference<>();
+        AtomicReference<Throwable> queryFailure = new AtomicReference<>();
+
+        Future<Void> queryFuture = executorService.submit(() -> {
+            try (Connection connection = createConnection("blackhole", "default")) {
+                try (Statement statement = connection.createStatement()) {
+                    try (ResultSet resultSet = statement.executeQuery("SELECT * FROM test_cancellation")) {
+                        queryId.set(resultSet.unwrap(PrestoResultSet.class).getQueryId());
+                        queryStarted.countDown();
+                        try {
+                            resultSet.next();
+                        }
+                        catch (Throwable t) {
+                            queryFailure.set(t);
+                        }
+                        finally {
+                            queryFinished.countDown();
+                        }
+                    }
+                }
+            }
+            return null;
+        });
+
+        queryStarted.await(1, MINUTES);
+        assertNotNull(queryId.get());
+        assertQueryState(queryId.get(), "QUEUED", "PLANNING", "STARTING", "RUNNING");
```

Comment:
This is brittle, since we might make changes to the state machine (and have recently, e.g. for transactions).

One fix is to use `QueryState.valueOf(...).isDone()`, but there's an easier way: select the `end` timestamp from the table. It will be null if the query is not done.